v2/svid/x509svid/svid.go

```diff
@@ -25,38 +32,145 @@ type SVID struct {
+		if errors.Is(err, pemutil.ErrUnexpectedBlockType) {
+			continue
+		}
+		if err != nil {
+			return nil, x509SVIDErr.New("cannot parse PEM encoded certificate: %v", err)
+		}
+		certs = append(certs, cert)
+	}
+
+	var privateKey crypto.PrivateKey
+	var err error
+	for {
+		if len(keyBytes) == 0 {
+			break
+		}
+		privateKey, keyBytes, err = pemutil.ParsePrivateKey(keyBytes)
+		if errors.Is(err, pemutil.ErrUnexpectedBlockType) {
+			continue
+		}
+		if err != nil {
+			return nil, x509SVIDErr.New("cannot parse PEM encoded private key: %v", err)
+		}
+		break
+	}
+
+	return newSVID(certs, privateKey)
 }
 
 // ParseRaw parses the X509-SVID from certificate and key bytes. The
 // certificate must be ASN.1 DER (concatenated with no intermediate
 // padding if there are more than one certificate). The key must be a PKCS#8
 // ASN.1 DER.
 func ParseRaw(certBytes, keyBytes []byte) (*SVID, error) {
-	panic("not implemented")
+	certificates, err := x509.ParseCertificates(certBytes)
+	if err != nil {
+		return nil, x509SVIDErr.New("cannot parse DER encoded certificate: %v", err)
+	}
+
+	privateKey, err := x509.ParsePKCS8PrivateKey(keyBytes)
+	if err != nil {
+		return nil, x509SVIDErr.New("cannot parse DER encoded private key: %v", err)
+	}
+
+	return newSVID(certificates, privateKey)
 }
 
 // Marshal marshals the X509-SVID and returns PEM encoded blocks for the SVID
 // and private key.
 func (s *SVID) Marshal() ([]byte, []byte, error) {
-	panic("not implemented")
+	certBytes := pemutil.EncodeCertificates(s.Certificates)
+	keyBytes, err := pemutil.EncodePKCS8PrivateKey(s.PrivateKey)
+	if err != nil {
+		return nil, nil, x509SVIDErr.New("cannot encode private key: %v", err)
+	}
+
+	return certBytes, keyBytes, nil
 }
 
 // MarshalRaw marshals the X509-SVID and returns ASN.1 DER for the certificates
 // (concatenated with no intermediate padding) and PKCS8 ASN1.DER for the
 // private key.
 func (s *SVID) MarshalRaw() ([]byte, []byte, error) {
-	panic("not implemented")
+	key, err := x509.MarshalPKCS8PrivateKey(s.PrivateKey)
+	if err != nil {
+		return nil, nil, x509SVIDErr.New("cannot marshal private key: %v", err)
+	}
+
+	certBytes := []byte{}
+	for _, cert := range s.Certificates {
+		certBytes = append(certBytes, cert.Raw...)
+	}
+
+	return certBytes, key, nil
 }
 
 // GetX509SVID returns the X509-SVID. It implements the Source interface.
 func (s *SVID) GetX509SVID() (*SVID, error) {
 	return s, nil
 }
+
+func newSVID(certificates []*x509.Certificate, privateKey interface{}) (*SVID, error) {
+	if len(certificates) == 0 {
+		return nil, x509SVIDErr.New("no certificates found")
+	}
+
+	spiffeID, err := getSPIFFEID(certificates[0])
+	if err != nil {
+		return nil, x509SVIDErr.New("cannot get SPIFFE ID: %v", err)
+	}
+
+	if privateKey == nil {
+		return nil, x509SVIDErr.New("no private key found")
+	}
+
+	signer, ok := privateKey.(crypto.Signer)
+	if !ok {
+		return nil, x509SVIDErr.New("expected crypto.Signer; got %T", privateKey)
+	}
+
+	return &SVID{
```

Comment:
Should we validate the cert/key after parsing to make sure it adjust to the spiffe standard for [x509-SVID](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md#2-spiffe-id)?

Not sure that is expected to be done here or in the `Verify`/`ParseVerify` package functions

Reply:
Yes, for sure we should ensure this is a valid SVID on ingress. We can add that after we merge the verification PR though.